### PR TITLE
[docs][tools] Docs and scripts for air-gapped registry

### PR DESCRIPTION
**Pull request by konstantin-axenov**
Signed-off-by: Konstantin Aksenov <konstantin.aksenov@flant.com>

## Description

-  add scripts to pull Deckhouse release images locally and push these images to air-gapped registry
-  add docs for this feature

## Why do we need it, and what problem does it solve?

There are conditions when clients can't use proxy registries because there is no access to Internet.

## Changelog entries

```changes
section: docs
type: feature
summary: Add docs for an air-gapped registry
impact_level: low
```

```changes
section: tools
type: feature
summary: Add scripts for an air-gapped registry.
```

### CODE REVIEW

File: docs/documentation/pages/DECKHOUSE-FAQ.md [hunk 1]
```diff
@@ -156,6 +156,46 @@ You need to use the Proxy Cache feature of a [Harbor](https://github.com/goharbo
...
+```shell
+curl -fsSL -o d8-pull.sh https://raw.githubusercontent.com/deckhouse/deckhouse/main/tools/release/d8-pull.sh
+chmod 700 d8-pull.sh
+./d8-pull.sh --license YOUR_DECKHOUSE_LICENSE_KEY --output-dir /your/output-dir/
+```
+
+Example of pulling images
+
+```shell
+./d8-push.sh --source-dir /your/source-dir/ --path your.private.registry.com/deckhouse --username YOUR_USERNAME --password YOUR_PASSWORD
```
Comment (z9r5) [inline docs/documentation/pages/DECKHOUSE-FAQ.md:172]:
```suggestion
./d8-pull.sh --source-dir /your/source-dir/ --path your.private.registry.com/deckhouse --username YOUR_USERNAME --password YOUR_PASSWORD
```

I'm not sure, but the d8-pull.sh script was downloaded.

File: docs/documentation/pages/DECKHOUSE-FAQ.md [hunk 2]
```diff
@@ -156,6 +156,46 @@ You need to use the Proxy Cache feature of a [Harbor](https://github.com/goharbo
...
+
+Download script on a host that have access to `registry.deckhouse.io`. `Docker`, `crane` and `jq` must be installed on this host.
+
+```shell
+curl -fsSL -o d8-pull.sh https://raw.githubusercontent.com/deckhouse/deckhouse/main/tools/release/d8-pull.sh
+chmod 700 d8-pull.sh
+./d8-pull.sh --license YOUR_DECKHOUSE_LICENSE_KEY --output-dir /your/output-dir/
+```
+
+Example of pulling images
```
Comment (z9r5) [inline docs/documentation/pages/DECKHOUSE-FAQ.md:169]:
```suggestion
Example of pulling images:
```

File: docs/documentation/pages/DECKHOUSE-FAQ.md [hunk 3]
```diff
@@ -156,6 +156,46 @@ You need to use the Proxy Cache feature of a [Harbor](https://github.com/goharbo
...
+
+```shell
+./d8-push.sh --source-dir /your/source-dir/ --path your.private.registry.com/deckhouse --username YOUR_USERNAME --password YOUR_PASSWORD
+```
+
+Upload folder from previous step to a host with access to an air-gapped registry. `Crane` must be installed on this host.
+Download script on a host that have access to an air-gapped registry.
```
Comment (z9r5) [inline docs/documentation/pages/DECKHOUSE-FAQ.md:176]:
```suggestion
Upload the folder from the previous step to a host with access to an air-gapped registry. `Crane` must be installed on this host.
Download script on a host that has access to an air-gapped registry.
```

File: docs/documentation/pages/DECKHOUSE-FAQ.md [hunk 4]
```diff
@@ -156,6 +156,46 @@ You need to use the Proxy Cache feature of a [Harbor](https://github.com/goharbo
...
+
+```shell
+curl -fsSL -o d8-push.sh https://raw.githubusercontent.com/deckhouse/deckhouse/main/tools/release/d8-push.sh
+chmod 700 d8-push.sh
+.d8-push.sh --source-dir /your/source-dir/ --path your.private.registry.com/deckhouse --username YOUR_USERNAME --password YOUR_PASSWORD
+```
+
+Example of pushing images
```
Comment (z9r5) [inline docs/documentation/pages/DECKHOUSE-FAQ.md:184]:
```suggestion
Example of pushing images:
```

File: docs/documentation/pages/DECKHOUSE-FAQ.md [hunk 5]
```diff
@@ -156,6 +156,46 @@ You need to use the Proxy Cache feature of a [Harbor](https://github.com/goharbo
...
+
+```shell
+./d8-push.sh --source-dir /your/source-dir/ --path your.private.registry.com/deckhouse --username YOUR_USERNAME --password YOUR_PASSWORD
+```
+
+## How to bootstrap cluster and run Deckhouse without usage of release channels?
```
Comment (z9r5) [inline docs/documentation/pages/DECKHOUSE-FAQ.md:190]:
```suggestion
## How to bootstrap a cluster and run Deckhouse without the usage of release channels?
```

File: docs/documentation/pages/DECKHOUSE-FAQ.md [hunk 6]
```diff
@@ -156,6 +156,46 @@ You need to use the Proxy Cache feature of a [Harbor](https://github.com/goharbo
...
+
+This case is only valid if you don't have release channel images in your air-gapped registry.
+
+* If you want to bootstrap cluster, you have to use exact tag of a Docker image to install the Deckhouse Platform.
```
Comment (z9r5) [inline docs/documentation/pages/DECKHOUSE-FAQ.md:194]:
```suggestion
* If you want to bootstrap a cluster, you have to use the exact tag of a Docker image to install the Deckhouse Platform.
```

File: docs/documentation/pages/DECKHOUSE-FAQ.md [hunk 7]
```diff
@@ -156,6 +156,46 @@ You need to use the Proxy Cache feature of a [Harbor](https://github.com/goharbo
...
+For example, if you want to install release v1.32.13, you have to use image `your.private.registry.com/deckhouse/install:v1.32.13`. And you have to use `devBranch: v1.32.13` instead of `releaseChannel: XXX` in `config.yml`.
+* If you already have cluster up and running, you have to remove `releaseChannel` setting from configmap `d8-system/deckhouse` and set appropriate `image` for `d8-system/deckhouse` deployment. Further updates have to be done by manual update of `image`.
```
Comment (z9r5) [inline docs/documentation/pages/DECKHOUSE-FAQ.md:196]:
```suggestion
* If you already have a cluster up and running, you have to remove `releaseChannel` setting from the ConfigMap `d8-system/deckhouse` and set the appropriate `image` parameter for `d8-system/deckhouse` Deployment. Further updates have to be done by a manual update of the `image` parameter of the `d8-system/deckhouse` Deployment.
```

File: docs/documentation/pages/DECKHOUSE-FAQ_RU.md [hunk 8]
```diff
@@ -160,6 +160,44 @@ deckhouse: |
...
+### Ручная загрузка образов в изолированный приватный registry
+
+Загрузите скрипт на хост, с которого есть доступ до `registry.deckhouse.io`. `Docker`, `crane` и `jq` должны быть установлены на хосте.
+
+```shell
+curl -fsSL -o d8-pull.sh https://raw.githubusercontent.com/deckhouse/deckhouse/main/tools/release/d8-pull.sh
+chmod 700 d8-pull.sh
+```
+
+Пример вызова команды скачивания образов
```
Comment (z9r5) [inline docs/documentation/pages/DECKHOUSE-FAQ_RU.md:172]:
```suggestion
Пример вызова команды скачивания образов:
```

File: docs/documentation/pages/DECKHOUSE-FAQ_RU.md [hunk 9]
```diff
@@ -160,6 +160,44 @@ deckhouse: |
...
+
+```shell
+./d8-pull.sh --license YOUR_DECKHOUSE_LICENSE_KEY --output-dir /your/output-dir/
+```
+
+Загрузите директорию с образами, полученную на предыдущем шаге, на хост, с которого есть доступ до изолированного приватного registry. `Crane` должен быть установлен на хосте.
+Загрузите скрипт на этот хост
```
Comment (z9r5) [inline docs/documentation/pages/DECKHOUSE-FAQ_RU.md:179]:
```suggestion
Загрузите скрипт на этот хост:
```

File: docs/documentation/pages/DECKHOUSE-FAQ_RU.md [hunk 10]
```diff
@@ -160,6 +160,44 @@ deckhouse: |
...
+
+```shell
+curl -fsSL -o d8-push.sh https://raw.githubusercontent.com/deckhouse/deckhouse/main/tools/release/d8-push.sh
+chmod 700 d8-push.sh
+```
+
+Пример вызова команды загрузки образов в изолированный приватный registry
```
Comment (z9r5) [inline docs/documentation/pages/DECKHOUSE-FAQ_RU.md:186]:
```suggestion
Пример вызова команды загрузки образов в изолированный приватный registry:
```

File: tools/release/d8-pull.sh [hunk 11]
```diff
@@ -0,0 +1,183 @@
+#!/bin/bash
+
+# Copyright 2021 Flant JSC
```
Comment (z9r5) [inline tools/release/d8-pull.sh:3]:
```suggestion
# Copyright 2022 Flant JSC
```

File: tools/release/d8-pull.sh [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+    return 1
+  fi
+
+  if [[ "$OUTPUT_DIR" == "" ]]; then
+    echo "--output-dir is required"
+    return 1
+  fi
+
+  if [[ "$EDITION" == "ee" ]] && [[ "$LICENSE" == "" ]]; then
+    echo "License is required to download ee edition of Deckhouse. Please provide it with cli argument --license."
```
Comment (z9r5) [inline tools/release/d8-pull.sh:129]:
```suggestion
    echo "License is required to download Deckhouse Enterprise Edition. Please provide it with CLI argument --license."
```

File: tools/release/d8-push.sh [hunk 13]
```diff
@@ -0,0 +1,128 @@
+#!/bin/bash
+
+# Copyright 2021 Flant JSC
```
Comment (z9r5) [inline tools/release/d8-push.sh:3]:
```suggestion
# Copyright 2022 Flant JSC
```

**Review by z9r5 — APPROVED** [review]

File: docs/documentation/pages/DECKHOUSE-FAQ.md (hunk 1 above)
Comment (konstantin-axenov) [reply docs/documentation/pages/DECKHOUSE-FAQ.md:172]: You're right, but not only command is wrong. Will fix it in separate commit.

**Review by RomanenkoDenys — APPROVED** [review]

**Review by nabokihms — APPROVED** [review]

**Pull request merged**